File: resource-tools/index-builder.js
#!/usr/bin/env node

/**
 * Resource Index Builder
 * Scans .claude/skills and .github/awesome-copilot-main to create lightweight indexes
 * This reduces AI token costs by providing metadata without full content
 */

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.resolve(__dirname, '..');
const OUTPUT_DIR = path.join(ROOT_DIR, 'resource-tools', 'indexes');

// Ensure output directory exists
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

/**
 * Parse frontmatter from markdown files
 */
function parseFrontmatter(content) {
  const frontmatterRegex = /^---\s*\n([\s\S]*?)\n---/;
  const match = content.match(frontmatterRegex);

  if (!match) return {};

  const frontmatter = {};
  const lines = match[1].split('\n');

  for (const line of lines) {
    const colonIndex = line.indexOf(':');
    if (colonIndex === -1) continue;

    const key = line.slice(0, colonIndex).trim();
    let value = line.slice(colonIndex + 1).trim();

    // Remove quotes if present
    if ((value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }

    frontmatter[key] = value;
  }

  return frontmatter;
}

/**
 * Extract first paragraph after frontmatter as summary
 */
function extractSummary(content, maxLength = 200) {
  // Remove frontmatter
  const withoutFrontmatter = content.replace(/^---\s*\n[\s\S]*?\n---\s*\n/, '');

  // Get first non-empty paragraph
  const paragraphs = withoutFrontmatter.split('\n\n');
  for (const para of paragraphs) {
    const cleaned = para.trim().replace(/^#+\s*/, ''); // Remove headers
    if (cleaned && !cleaned.startsWith('```')) {
      return cleaned.length > maxLength
        ? cleaned.slice(0, maxLength) + '...'
        : cleaned;
    }
  }

  return '';
}

/**
 * Index Claude skills from .claude/skills
 */
function indexClaudeSkills() {
  const skillsDir = path.join(ROOT_DIR, '.claude', 'skills');
  if (!fs.existsSync(skillsDir)) {
    console.log('No .claude/skills directory found');
    return [];
  }

  const skills = [];
  const skillDirs = fs.readdirSync(skillsDir);

  for (const skillName of skillDirs) {
    const skillPath = path.join(skillsDir, skillName);
    const skillFile = path.join(skillPath, 'SKILL.md');

    if (!fs.existsSync(skillFile)) continue;

    const content = fs.readFileSync(skillFile, 'utf-8');
    const frontmatter = parseFrontmatter(content);
    const summary = extractSummary(content);

    skills.push({
      name: frontmatter.name || skillName,
      description: frontmatter.description || summary,
      type: 'claude-skill',
      path: path.relative(ROOT_DIR, skillFile),
      tags: frontmatter.tags ? frontmatter.tags.split(',').map(t => t.trim()) : [],
      size: Buffer.byteLength(content, 'utf-8')
    });
  }

  return skills;
}

/**
 * Index agents from .github/awesome-copilot-main/agents
 */
function indexAgents() {
  const agentsDir = path.join(ROOT_DIR, '.github', 'awesome-copilot-main', 'agents');
  if (!fs.existsSync(agentsDir)) {
    console.log('No agents directory found');
    return [];
  }

  const agents = [];
  const agentFiles = fs.readdirSync(agentsDir).filter(f => f.endsWith('.agent.md'));

  for (const agentFile of agentFiles) {
    const agentPath = path.join(agentsDir, agentFile);
    const content = fs.readFileSync(agentPath, 'utf-8');
    const frontmatter = parseFrontmatter(content);

    agents.push({
      name: frontmatter.name || path.basename(agentFile, '.agent.md'),
      description: frontmatter.description || '',
      type: 'copilot-agent',
      path: path.relative(ROOT_DIR, agentPath),
      tools: frontmatter.tools ? JSON.parse(frontmatter.tools.replace(/'/g, '"')) : [],
      size: Buffer.byteLength(content, 'utf-8')
    });
  }

  return agents;
}

/**
 * Index prompts from .github/awesome-copilot-main/prompts
 */
function indexPrompts() {
  const promptsDir = path.join(ROOT_DIR, '.github', 'awesome-copilot-main', 'prompts');
  if (!fs.existsSync(promptsDir)) {
    console.log('No prompts directory found');
    return [];
  }

  const prompts = [];
  const promptFiles = fs.readdirSync(promptsDir).filter(f => f.endsWith('.prompt.md'));

  for (const promptFile of promptFiles) {
    const promptPath = path.join(promptsDir, promptFile);
    const content = fs.readFileSync(promptPath, 'utf-8');
    const frontmatter = parseFrontmatter(content);

    prompts.push({
      name: frontmatter.name || path.basename(promptFile, '.prompt.md'),
      description: frontmatter.description || '',
      type: 'copilot-prompt',
      path: path.relative(ROOT_DIR, promptPath),
      tags: frontmatter.tags ? frontmatter.tags.split(',').map(t => t.trim()) : [],
      size: Buffer.byteLength(content, 'utf-8')
    });
  }

  return prompts;
}

/**
 * Index instructions from .github/awesome-copilot-main/instructions
 */
function indexInstructions() {
  const instructionsDir = path.join(ROOT_DIR, '.github', 'awesome-copilot-main', 'instructions');
  if (!fs.existsSync(instructionsDir)) {
    console.log('No instructions directory found');
    return [];
  }

  const instructions = [];
  const instructionFiles = fs.readdirSync(instructionsDir).filter(f => f.endsWith('.instruction.md'));

  for (const instructionFile of instructionFiles) {
    const instructionPath = path.join(instructionsDir, instructionFile);
    const content = fs.readFileSync(instructionPath, 'utf-8');
    const frontmatter = parseFrontmatter(content);

    instructions.push({
      name: frontmatter.name || path.basename(instructionFile, '.instruction.md'),
      description: frontmatter.description || '',
      type: 'copilot-instruction',
      path: path.relative(ROOT_DIR, instructionPath),
      appliesTo: frontmatter.appliesTo || frontmatter.patterns || '',
      size: Buffer.byteLength(content, 'utf-8')
    });
  }

  return instructions;
}

/**
 * Build master index
 */
function buildIndex() {
  console.log('Building resource index...\n');

  const skills = indexClaudeSkills();
  console.log(`✓ Indexed ${skills.length} Claude skills`);

  const agents = indexAgents();
  console.log(`✓ Indexed ${agents.length} Copilot agents`);

  const prompts = indexPrompts();
  console.log(`✓ Indexed ${prompts.length} Copilot prompts`);

  const instructions = indexInstructions();
  console.log(`✓ Indexed ${instructions.length} Copilot instructions`);

  const masterIndex = {
    generatedAt: new Date().toISOString(),
    stats: {
      totalResources: skills.length + agents.length + prompts.length + instructions.length,
      skills: skills.length,
      agents: agents.length,
      prompts: prompts.length,
      instructions: instructions.length
    },
    resources: {
      skills,
      agents,
      prompts,
      instructions
    }
  };

  // Write master index
  const indexPath = path.join(OUTPUT_DIR, 'master-index.json');
  fs.writeFileSync(indexPath, JSON.stringify(masterIndex, null, 2));
  console.log(`\n✓ Master index written to ${path.relative(ROOT_DIR, indexPath)}`);

  // Write separate indexes for faster lookups
  fs.writeFileSync(path.join(OUTPUT_DIR, 'skills-index.json'), JSON.stringify(skills, null, 2));
  fs.writeFileSync(path.join(OUTPUT_DIR, 'agents-index.json'), JSON.stringify(agents, null, 2));
  fs.writeFileSync(path.join(OUTPUT_DIR, 'prompts-index.json'), JSON.stringify(prompts, null, 2));
  fs.writeFileSync(path.join(OUTPUT_DIR, 'instructions-index.json'), JSON.stringify(instructions, null, 2));

  console.log('✓ Separate indexes written\n');

  // Print statistics
  const totalSize = [...skills, ...agents, ...prompts, ...instructions]
    .reduce((sum, r) => sum + r.size, 0);

  console.log('Statistics:');
  console.log(`  Total resources: ${masterIndex.stats.totalResources}`);
  console.log(`  Total content size: ${(totalSize / 1024).toFixed(2)} KB`);
  console.log(`  Index size: ${(Buffer.byteLength(JSON.stringify(masterIndex), 'utf-8') / 1024).toFixed(2)} KB`);
  console.log(`  Space savings: ${((1 - (Buffer.byteLength(JSON.stringify(masterIndex), 'utf-8') / totalSize)) * 100).toFixed(1)}%`);
}

// Run the indexer
buildIndex();

File: resource-tools/index_builder.py
#!/usr/bin/env python3

"""
Resource Index Builder
Scans .claude/skills and _reference/awesome-copilot to create lightweight indexes
This reduces AI token costs by providing metadata without full content
"""

import json
import os
import re
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Optional

WORKSPACE_DIR = Path(__file__).parent.parent  # E:\_Development
REFERENCE_DIR = WORKSPACE_DIR / '_reference' / 'awesome-copilot'
OUTPUT_DIR = Path(__file__).parent / 'indexes'

# Ensure output directory exists
OUTPUT_DIR.mkdir(exist_ok=True)


def parse_frontmatter(content: str) -> Dict[str, Any]:
    """Parse YAML frontmatter from markdown files"""
    frontmatter_pattern = r'^---\s*\n(.*?)\n---'
    match = re.match(frontmatter_pattern, content, re.DOTALL)

    if not match:
        return {}

    frontmatter = {}
    lines = match.group(1).split('\n')

    for line in lines:
        if ':' not in line:
            continue

        key, value = line.split(':', 1)
        key = key.strip()
        value = value.strip()

        # Remove quotes if present
        if (value.startswith('"') and value.endswith('"')) or \
           (value.startswith("'") and value.endswith("'")):
            value = value[1:-1]

        frontmatter[key] = value

    return frontmatter


def extract_summary(content: str, max_length: int = 200) -> str:
    """Extract first paragraph after frontmatter as summary"""
    # Remove frontmatter
    without_frontmatter = re.sub(r'^---\s*\n.*?\n---\s*\n', '', content, flags=re.DOTALL)

    # Get first non-empty paragraph
    paragraphs = without_frontmatter.split('\n\n')
    for para in paragraphs:
        cleaned = para.strip().lstrip('#').strip()  # Remove headers
        if cleaned and not cleaned.startswith('```'):
            return cleaned[:max_length] + '...' if len(cleaned) > max_length else cleaned

    return ''


def index_claude_skills() -> List[Dict[str, Any]]:
    """Index Claude skills from .claude/skills"""
    skills_dir = WORKSPACE_DIR / '.claude' / 'skills'
    if not skills_dir.exists():
        print('No .claude/skills directory found')
        return []

    skills = []

    for skill_dir in skills_dir.iterdir():
        if not skill_dir.is_dir():
            continue

        skill_file = skill_dir / 'SKILL.md'
        if not skill_file.exists():
            continue

        try:
            content = skill_file.read_text(encoding='utf-8')
            frontmatter = parse_frontmatter(content)
            summary = extract_summary(content)

            skills.append({
                'name': frontmatter.get('name', skill_dir.name),
                'description': frontmatter.get('description', summary),
                'type': 'claude-skill',
                'path': str(skill_file.relative_to(WORKSPACE_DIR)).replace('\\', '/'),
                'tags': [t.strip() for t in frontmatter.get('tags', '').split(',')] if frontmatter.get('tags') else [],
                'size': len(content.encode('utf-8'))
            })
        except Exception as e:
            print(f'Error processing {skill_file}: {e}')

    return skills


def index_agents() -> List[Dict[str, Any]]:
    """Index agents from .github/awesome-copilot-main/agents"""
    agents_dir = REFERENCE_DIR / 'agents'
    if not agents_dir.exists():
        print('No agents directory found')
        return []

    agents = []

    for agent_file in agents_dir.glob('*.agent.md'):
        try:
            content = agent_file.read_text(encoding='utf-8')
            frontmatter = parse_frontmatter(content)

            # Parse tools array
            tools = []
            if 'tools' in frontmatter:
                tools_str = frontmatter['tools'].strip()
                if tools_str.startswith('[') and tools_str.endswith(']'):
                    # Parse JSON-like array
                    tools_str = tools_str.replace("'", '"')
                    try:
                        tools = json.loads(tools_str)
                    except:
                        # Fallback to simple split
                        tools = [t.strip().strip('"').strip("'") for t in tools_str[1:-1].split(',')]

            agents.append({
                'name': frontmatter.get('name', agent_file.stem.replace('.agent', '')),
                'description': frontmatter.get('description', ''),
                'type': 'copilot-agent',
                'path': str(agent_file.relative_to(WORKSPACE_DIR)).replace('\\', '/'),
                'tools': tools,
                'size': len(content.encode('utf-8'))
            })
        except Exception as e:
            print(f'Error processing {agent_file}: {e}')

    return agents


def index_prompts() -> List[Dict[str, Any]]:
    """Index prompts from .github/awesome-copilot-main/prompts"""
    prompts_dir = REFERENCE_DIR / 'prompts'
    if not prompts_dir.exists():
        print('No prompts directory found')
        return []

    prompts = []

    for prompt_file in prompts_dir.glob('*.prompt.md'):
        try:
            content = prompt_file.read_text(encoding='utf-8')
            frontmatter = parse_frontmatter(content)

            prompts.append({
                'name': frontmatter.get('name', prompt_file.stem.replace('.prompt', '')),
                'description': frontmatter.get('description', ''),
                'type': 'copilot-prompt',
                'path': str(prompt_file.relative_to(WORKSPACE_DIR)).replace('\\', '/'),
                'tags': [t.strip() for t in frontmatter.get('tags', '').split(',')] if frontmatter.get('tags') else [],
                'size': len(content.encode('utf-8'))
            })
        except Exception as e:
            print(f'Error processing {prompt_file}: {e}')

    return prompts


def index_instructions() -> List[Dict[str, Any]]:
    """Index instructions from .github/awesome-copilot-main/instructions"""
    instructions_dir = REFERENCE_DIR / 'instructions'
    if not instructions_dir.exists():
        print('No instructions directory found')
        return []

    instructions = []

    for instruction_file in instructions_dir.glob('*.instruction.md'):
        try:
            content = instruction_file.read_text(encoding='utf-8')
            frontmatter = parse_frontmatter(content)

            instructions.append({
                'name': frontmatter.get('name', instruction_file.stem.replace('.instruction', '')),
                'description': frontmatter.get('description', ''),
                'type': 'copilot-instruction',
                'path': str(instruction_file.relative_to(WORKSPACE_DIR)).replace('\\', '/'),
                'appliesTo': frontmatter.get('appliesTo', frontmatter.get('patterns', '')),
                'size': len(content.encode('utf-8'))
            })
        except Exception as e:
            print(f'Error processing {instruction_file}: {e}')

    return instructions


def build_index():
    """Build master index"""
    print('Building resource index...\n')

    skills = index_claude_skills()
    print(f'[OK] Indexed {len(skills)} Claude skills')

    agents = index_agents()
    print(f'[OK] Indexed {len(agents)} Copilot agents')

    prompts = index_prompts()
    print(f'[OK] Indexed {len(prompts)} Copilot prompts')

    instructions = index_instructions()
    print(f'[OK] Indexed {len(instructions)} Copilot instructions')

    master_index = {
        'generatedAt': datetime.utcnow().isoformat() + 'Z',
        'stats': {
            'totalResources': len(skills) + len(agents) + len(prompts) + len(instructions),
            'skills': len(skills),
            'agents': len(agents),
            'prompts': len(prompts),
            'instructions': len(instructions)
        },
        'resources': {
            'skills': skills,
            'agents': agents,
            'prompts': prompts,
            'instructions': instructions
        }
    }

    # Write master index
    index_path = OUTPUT_DIR / 'master-index.json'
    with open(index_path, 'w', encoding='utf-8') as f:
        json.dump(master_index, f, indent=2)
    print(f'\n[OK] Master index written to {index_path.relative_to(WORKSPACE_DIR)}')

    # Write separate indexes for faster lookups
    (OUTPUT_DIR / 'skills-index.json').write_text(json.dumps(skills, indent=2), encoding='utf-8')
    (OUTPUT_DIR / 'agents-index.json').write_text(json.dumps(agents, indent=2), encoding='utf-8')
    (OUTPUT_DIR / 'prompts-index.json').write_text(json.dumps(prompts, indent=2), encoding='utf-8')
    (OUTPUT_DIR / 'instructions-index.json').write_text(json.dumps(instructions, indent=2), encoding='utf-8')

    print('[OK] Separate indexes written\n')

    # Print statistics
    total_size = sum(r['size'] for r in skills + agents + prompts + instructions)
    index_size = len(json.dumps(master_index).encode('utf-8'))

    print('Statistics:')
    print(f'  Total resources: {master_index["stats"]["totalResources"]}')
    print(f'  Total content size: {total_size / 1024:.2f} KB')
    print(f'  Index size: {index_size / 1024:.2f} KB')
    print(f'  Space savings: {(1 - index_size / total_size) * 100:.1f}%')


if __name__ == '__main__':
    build_index()
